From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pravalika Gurram <pravalika.gurram@intel.com>
Date: Mon, 13 Jul 2026 19:44:23 +0530
Subject: [PATCH] drm/pagemap: Use zone_device_page_init() for kernels < v6.19

zone_device_folio_init() was introduced in kernel v6.19. For older kernels,
fall back to zone_device_page_init() for zone device memory initialization.

Limitations of Fallback:
- No support for large folios (2MB, 1GB, etc.)
- Cannot efficiently handle 2MB+ contiguous zone device allocations
- Loss of huge page benefits and THP support

Reference:
	d245f9b4ab80 mm/zone_device: support large zone device private folios

Signed-off-by: Pravalika Gurram <pravalika.gurram@intel.com>
---
 drivers/gpu/drm/drm_pagemap.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/gpu/drm/drm_pagemap.c b/drivers/gpu/drm/drm_pagemap.c
index 5002049..fbd7c48 100644
--- a/drivers/gpu/drm/drm_pagemap.c
+++ b/drivers/gpu/drm/drm_pagemap.c
@@ -210,9 +210,14 @@ static void drm_pagemap_get_devmem_page(struct page *page,
 					unsigned int order,
 					struct drm_pagemap_zdd *zdd)
 {
+#ifdef BPM_ZONE_DEVICE_FOLIO_INIT_NOT_PRESENT
+	page->zone_device_data = drm_pagemap_zdd_get(zdd);
+	zone_device_page_init(page, page_pgmap(page), 0);
+#else
 	zone_device_folio_init((struct folio *)page, zdd->dpagemap->pagemap,
 			       order);
 	folio_set_zone_device_data(page_folio(page), drm_pagemap_zdd_get(zdd));
+#endif
 }
 
 /**
-- 
2.43.0

